src/pytest_bdd/parser.py

```diff
@@ -454,6 +491,22 @@ def parse(self) -> Feature:
         for child in feature_data.children:
             if child.background:
                 feature.background = self.parse_background(child.background, feature)
+            if child.rule:
+                _backgrounds = [c.background for c in child.rule.children if c.background]
+                _rule = Rule(
+                    keyword=child.rule.keyword,
+                    name=child.rule.name,
+                    description=child.rule.description,
+                    tags=self.get_tag_names(child.rule.tags),
+                    feature=feature,
+                )
+                if _backgrounds:
+                    _rule.add_background(self.parse_background(_backgrounds[0], _rule))
+                rule_scenarios = [c.scenario for c in child.rule.children if c.scenario]
+                if rule_scenarios:
+                    for rule_scenario in rule_scenarios:
+                        scenario = self.parse_scenario(rule_scenario, feature, _rule)
+                        feature.scenarios[scenario.name] = scenario
```

Comment:
I wonder if there's a way to make this logic more straightforward.
There is a lot of mutations and it's hard to keep track of what's going on....

Reply:
Extracted the logic into methods to make easier to read